### PR TITLE
JenkinsFile: Build on server

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
This is a big update and cleanup of the JenkinsFile.

The notable changes are:
- Reproducible builds with `riot_build` build server that runs `BUILD_IN_DOCKER`
- Extra parameters for more control over what is tested and how to test
- cleanup and documentation of functions
- Use gitcache to reduce network traffic

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -75,7 +101,163 @@ void createPipelineTriggers() {
...
+def getUrlAndCommitFromParams(ver, repo_owner, repo_name, pull = 0) {
+    if (ver == 'master') {
+        return getUrlAndCommitFromDefault(repo_owner, repo_name)
+    }
+    else if (ver == 'pull') {
+        assert pull != '0' : "Selected pull of ${repo_name} but no pull entered"
+        return getUrlAndCommitFromPr(repo_owner, repo_name, pull)
+    }
+    else if (ver == 'current') {
+        return ["https://github.com/${repo_owner}/${repo_name}",
```
Comment (pokgak) [inline Jenkinsfile:148]:
```suggestion
        return ["https://github.com/RIOT-OS/RobotFW-tests",
```

Might as well specify rf repo directly as we are reading `env.GIT_COMMIT` which is the current commit in rf repo. In 'submodule' we also use rf directly, so no problem this being too specific I guess.

File: Jenkinsfile [hunk 2]
```diff
@@ -75,7 +101,163 @@ void createPipelineTriggers() {
...
+ */
+def getUrlAndCommitFromDefault(repo_owner, repo_name) {
+    def query = "-X POST -d \"{\\\"query\\\": \\\"query "
+    query = "${query}{repository(name: \\\\\\\"${repo_name}\\\\\\\","
+    query = "${query}owner: \\\\\\\"${repo_owner}\\\\\\\") "
+    query = "${query}{defaultBranchRef "
+    query = "${query}{ target { oid } }}}\\\"}\" "
+    query = "${query}https://api.github.com/graphql"
+    def res = ""
+    withCredentials([string(credentialsId: 'mrkevinweiss_github_api_token',
```
Comment (pokgak) [inline Jenkinsfile:172]: Is the credentialsId defined in Jenkins manually? Might be better if we have a dedicated api token for Jenkins.

File: Jenkinsfile [hunk 3]
```diff
@@ -75,7 +101,163 @@ void createPipelineTriggers() {
...
+            variable: 'TOKEN')]) {
+        res = sh script: '''
+        curl -H "Authorization: token $TOKEN" ''' + query,
+        label: "Fetching commit id", returnStdout: true
+    }
```
Comment (pokgak) [inline Jenkinsfile:177]: this is used multiple times with different queries. maybe refactor it to a function accepting the query as the parameter?

File: Jenkinsfile [hunk 4]
```diff
@@ -75,7 +101,163 @@ void createPipelineTriggers() {
...
+ * @param repo_owner    The owner of the repo.
+ * @param repo_name     The name of the repo that contains the subrepo.
+ *
+ * @return              [url, commit_id]
+ */
+def getUrlAndCommitOfSubmodule(repo_owner, repo_name) {
+    def query = "-X POST -d \"{\\\"query\\\": \\\"query "
+    query = "${query}{repository(name: \\\\\\\"${repo_name}\\\\\\\","
+    query = "${query}owner: \\\\\\\"${repo_owner}\\\\\\\") "
+    query = "${query}{ submodules(first: 1) "
```
Comment (pokgak) [inline Jenkinsfile:244]: Rather than getting the first submodule in the repo, is it possible to get the repo by name instead?

File: Jenkinsfile [hunk 5]
```diff
@@ -88,116 +270,250 @@ def step_compile_results()
...
+ *
+ * For example, if a "board=samr21-xpro" and the test is "tests/periph_gpio",
+ * the binaries (.hex, .bin, .elf) will be stashed in
+ * "samr21_xpro_tests_periph_gpio".
+ *
+ * @param board The board to build
+ * @param test  The test to build
+ */
+def buildJob(board, test) {
+    exit_code = sh script: "${params.EXTRA_BUILD_ARGS} BUILD_IN_DOCKER=1 BOARD=${board} make -C ${test} clean all -j",
```
Comment (pokgak) [inline Jenkinsfile:420]: Note that there is the issue https://github.com/RIOT-OS/RIOT/issues/14504 when passing extra build args to docker. There is also a workaround linked there but I'm not sure how that applies here.

File: Jenkinsfile [hunk 6]
```diff
@@ -88,116 +270,250 @@ def step_compile_results()
...
+        returnStatus: true,
+        label: "Build BOARD=${board} TEST=${test}"
```
Comment (pokgak) [inline Jenkinsfile:422]:
```suggestion
        label: "Build ${params.EXTRA_BUIL_ARGS} BOARD=${board} TEST=${test}"
```

**Review by pokgak — CHANGES_REQUESTED** [review]
Looks good. Added some suggestions.

File: Jenkinsfile (hunk 2 above)
Comment (MrKevinWeiss) [reply Jenkinsfile:172]: Yup I forgot I left that in.

File: Jenkinsfile (hunk 4 above)
Comment (MrKevinWeiss) [reply Jenkinsfile:244]: It seems not

File: Jenkinsfile (hunk 5 above)
Comment (MrKevinWeiss) [reply Jenkinsfile:420]: I made it more generic, We also don't have the proper build server setup so there will probably be some tuning after the fact.

File: Jenkinsfile (hunk 6 above)
Comment (MrKevinWeiss) [reply Jenkinsfile:422]: You can see the call anyways, the label is just a quick view.

**Review by pokgak — CHANGES_REQUESTED** [review]

**Review by pokgak — APPROVED** [review]
LGTM. Ack.

**Pull request merged**